pkg/callback/scheduler_callback.go

```diff
@@ -118,15 +118,36 @@ func (callback *AsyncRMCallback) RecvUpdateResponse(response *si.UpdateResponse)
 	for _, release := range response.ReleasedAllocations {
 		log.Logger().Debug("callback: response to released allocations",
 			zap.String("UUID", release.UUID))
+
+		// delete the allocated pod
+		if app := callback.context.GetApplicationInternal(release.ApplicationID); app != nil {
+			// TerminationType 0 mean STOPPED_BY_RM
+			if release.TerminationType != si.AllocationRelease_STOPPED_BY_RM {
+				for _, task := range app.GetTaskMap() {
+					if task.GetTaskAllocationUUID() == release.UUID {
+						err := task.DeleteTaskPod(task.GetTaskPod())
+						if err != nil {
+							log.Logger().Error("failed to delete pod", zap.Error(err))
+						}
+					}
+				}
+			}
+		}
 	}
 
 	// handle status changes
 	for _, updated := range response.UpdatedApplications {
 		log.Logger().Debug("status update callback received",
 			zap.String("appId", updated.ApplicationID),
 			zap.String("new status", updated.State))
-
-		//handle status update
+		// delete application from context
+		if updated.State == events.States().Application.Completed {
+			err := callback.context.RemoveApplicationInternal(updated.ApplicationID)
+			if err != nil {
+				log.Logger().Error("failed to delete application", zap.Error(err))
+			}
+		}
+		// handle status update
 		dispatcher.Dispatch(cache.NewApplicationStatusChangeEvent(updated.ApplicationID, events.AppStateChange, updated.State))
```

Comment:
after we delete the app in line 145, it continues to handle the event in line 151. this seems to be problematic. 
should we add an `else` here?

Reply:
OK, done.